common/features.c

```diff
@@ -121,14 +121,26 @@ static const struct dependency feature_deps[] = {
 #endif
 };
 
-static void clear_feature_bit(u8 *features, u32 bit)
+static void clear_feature_bit(u8 **features, u32 bit)
 {
-	size_t bytenum = bit / 8, bitnum = bit % 8, len = tal_count(features);
+	size_t bytenum = bit / 8, bitnum = bit % 8, len = tal_count(*features),
+	       new_len;
 
 	if (bytenum >= len)
 		return;
 
-	features[len - 1 - bytenum] &= ~(1 << bitnum);
+	(*features)[len - 1 - bytenum] &= ~(1 << bitnum);
+
+	/* What is the new feature array length? */
+	new_len = len;
+	while (new_len > 0 && (*features)[len - new_len] == 0)
+		new_len--;
+
+	u8 *newarr = tal_arr(tal_parent(*features), u8, new_len);
+	if (new_len != 0)
+		memcpy(newarr, *features + (len - new_len), new_len);
+	tal_free(*features);
+	*features = newarr;
```

Comment:
This definitely confused me, because it looks like an open-coded tal_resize().

But ofc the features are big-endian, so we're stripping off the *front*:
```
/* Big-endian bitfields are weird, but it means we trim from the front: */
for (size_t trim = 0; trim < len; trim++) {
   if (features[trim])
       break;
}
memmove(*features, *features + trim, len - trim);
tal_resize(features, len - trim);
```

/* Trim features
memmove(*features, *features + (len - new_len), new_len)

Reply:
Would it make sense to extract this into a separate function `features_strip()` that minimizes the featurebit array as per spec? We're calling `clear_feature_bit` in a loop in the next commit potentially resizing the array on each iteration, which potentially results in intermediate allocations.

Reply:
I like this idea, of having it as a separate function. I do worry that it creates a 'non-obvious' dependency, in that every 'clear-feature-bit' *must* call `strip`. Guess it's not really a problem since it's all internal/private to this class, though...